### PR TITLE
add CinemaHall and MovieSession with implementations

### PR DESCRIPTION
**Pull request by KovalAlexey**

### CODE REVIEW

**Review by andriylagun — APPROVED** [review]

File: src/main/java/com/dev/cinema/service/impl/MovieSessionServiceImpl.java [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+import com.dev.cinema.lib.Service;
+import com.dev.cinema.model.MovieSession;
+import com.dev.cinema.service.MovieSessionService;
+import java.time.LocalDate;
+import java.util.List;
+
+@Service
+public class MovieSessionServiceImpl implements MovieSessionService {
+    @Inject
+    MovieSessionDao movieSessionDao;
```
Comment (RomanDubovskyi) [inline src/main/java/com/dev/cinema/service/impl/MovieSessionServiceImpl.java]: what about access modifier?

File: src/main/java/com/dev/cinema/service/impl/CinemaHallServiceImpl.java [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+import com.dev.cinema.lib.Inject;
+import com.dev.cinema.lib.Service;
+import com.dev.cinema.model.CinemaHall;
+import com.dev.cinema.service.CinemaHallService;
+import java.util.List;
+
+@Service
+public class CinemaHallServiceImpl implements CinemaHallService {
+    @Inject
+    CinemaHallDao cinemaHallDao;
```
Comment (RomanDubovskyi) [inline src/main/java/com/dev/cinema/service/impl/CinemaHallServiceImpl.java]: same

File: src/main/java/com/dev/cinema/dao/impl/MovieSessionDaoImpl.java [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+            session = HibernateUtil.getSessionFactory().openSession();
+            transaction = session.beginTransaction();
+            session.persist(movieSession);
+            transaction.commit();
+            return movieSession;
+        } catch (Exception e) {
+            if (transaction != null) {
+                transaction.rollback();
+            }
+            throw new DataProcesingException("Can't add new session", e);
```
Comment (RomanDubovskyi) [inline src/main/java/com/dev/cinema/dao/impl/MovieSessionDaoImpl.java]: let's make message more informative and add some info about entity we were trying to save (same goes for other methods with input parameters, where we throw exception)

File: src/main/java/com/dev/cinema/dao/impl/MovieSessionDaoImpl.java [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+    public List<MovieSession> findAvailableSessions(Long movieId, LocalDate date) {
+        LocalDateTime startOfDay = date.atStartOfDay();
+        LocalDateTime endOfDay = startOfDay.plusDays(1).minusSeconds(1);
+        try (Session session = HibernateUtil.getSessionFactory().openSession()) {
+            Query<MovieSession> query = session.createQuery("FROM MovieSession"
+                    + " WHERE movie_id =: movieId"
+                    + " AND showTime BETWEEN :startOfDay AND :endOfDay", MovieSession.class);
+            query.setParameter("movieId", movieId);
+            query.setParameter("startOfDay", startOfDay);
+            query.setParameter("endOfDay", endOfDay);
```
Comment (RomanDubovskyi) [inline src/main/java/com/dev/cinema/dao/impl/MovieSessionDaoImpl.java]:
```suggestion
            query.setParameter("startOfDay", date.atStartOfDat());
            query.setParameter("endOfDay", date.atTime(LocalTime.MAX);
```

File: src/main/java/com/dev/cinema/service/impl/MovieSessionServiceImpl.java (hunk 1 above)
Comment (KovalAlexey) [reply src/main/java/com/dev/cinema/service/impl/MovieSessionServiceImpl.java]: done

File: src/main/java/com/dev/cinema/service/impl/CinemaHallServiceImpl.java (hunk 2 above)
Comment (KovalAlexey) [reply src/main/java/com/dev/cinema/service/impl/CinemaHallServiceImpl.java]: done

File: src/main/java/com/dev/cinema/dao/impl/MovieSessionDaoImpl.java (hunk 4 above)
Comment (KovalAlexey) [reply src/main/java/com/dev/cinema/dao/impl/MovieSessionDaoImpl.java]: done

File: src/main/java/com/dev/cinema/dao/impl/MovieSessionDaoImpl.java (hunk 3 above)
Comment (KovalAlexey) [reply src/main/java/com/dev/cinema/dao/impl/MovieSessionDaoImpl.java]: done

File: src/main/java/com/dev/cinema/model/MovieSession.java [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+import javax.persistence.ManyToOne;
+import javax.persistence.Table;
+
+@Entity
+@Table(name = "movie_sessions")
+public class MovieSession {
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+    @ManyToOne
```
Comment (RomanDubovskyi) [inline src/main/java/com/dev/cinema/model/MovieSession.java:17]: this is for your info, but keep in mind that toOne relations have fetchType eager by default, which is not always quite good

File: src/main/java/com/dev/cinema/dao/impl/MovieSessionDaoImpl.java [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+            session = HibernateUtil.getSessionFactory().openSession();
+            transaction = session.beginTransaction();
+            session.persist(movieSession);
+            transaction.commit();
+            return movieSession;
+        } catch (Exception e) {
+            if (transaction != null) {
+                transaction.rollback();
+            }
+            throw new DataProcesingException("Can't add new session", e);
```
Comment (RomanDubovskyi) [inline src/main/java/com/dev/cinema/dao/impl/MovieSessionDaoImpl.java]: you've missed this message, don't forget to fix it

**Review by RomanDubovskyi — APPROVED** [review]

File: src/main/java/com/dev/cinema/dao/impl/MovieSessionDaoImpl.java (hunk 6 above)
Comment (KovalAlexey) [reply src/main/java/com/dev/cinema/dao/impl/MovieSessionDaoImpl.java]: done

File: src/main/java/com/dev/cinema/model/MovieSession.java (hunk 5 above)
Comment (KovalAlexey) [reply src/main/java/com/dev/cinema/model/MovieSession.java:17]: I know, I'll refactor this later with JOIN FETCH

**Pull request merged**